src/plugins/aws/idc/index.ts

```diff
@@ -0,0 +1,199 @@
+      );
+      return await response.json();
+    },
+    {
+      duration: 7.776e9, // the token has lifetime of 90 days
+    },
+    (data) => data.clientSecretExpiresAt < Date.now()
+  );
+
+// exchange the oidc response for aws credentials
+const exchangeForAWSCredentials = async (
+  oidcResponse: AWSTokenResponse,
+  idc: { id: string; region: string },
+  request: { account?: string; permissionSet: string }
+) => {
+  return await retryWithBackOff(
+    async () => {
+      const init = {
+        method: "GET",
+        headers: {
+          "x-amz-sso_bearer_token": oidcResponse.accessToken,
+        },
+      };
+      const { account, permissionSet } = request;
+      const { region } = idc;
+      const response = await fetch(
+        `https://portal.sso.${region}.amazonaws.com/federation/credentials?account_id=${account}&role_name=${permissionSet}`,
+        init
+      );
+      if (response.ok) return await response.json();
+      throw new Error(
+        `Failed to fetch AWS credentials: ${response.statusText}`
+      );
+    },
+    () => true,
+    3
+  );
+};
+
+const idcRequestBuilder = (
+  clientCredentials: { clientId: string; clientSecret: string },
+  idc: { id: string; region: string },
+  authorizeResponse: AWSAuthorizeResponse
+) => {
+  const { clientId, clientSecret } = clientCredentials;
+  const { region } = idc;
+
+  return {
+    url: `https://oidc.${region}.amazonaws.com/token`,
+    init: {
+      method: "POST",
+      body: JSON.stringify({
+        clientId,
+        clientSecret,
+        deviceCode: authorizeResponse.deviceCode,
+        grantType: DEVICE_GRANT_TYPE,
+      }),
+    },
+  };
+};
+
+const idcAuthorizeRequestBuilder = (
+  clientCredentials: { clientId: string; clientSecret: string },
+  idc: { id: string; region: string }
+) => {
+  const { clientId, clientSecret } = clientCredentials;
+  const { id, region } = idc;
+
+  return {
+    init: {
+      method: "POST",
+      body: JSON.stringify({
+        clientId,
+        clientSecret,
+        startUrl: `https://${id}.awsapps.com/start`,
+      }),
+    },
+    url: `https://oidc.${region}.amazonaws.com/device_authorization`,
+  };
+};
+
+export const idcLoginSteps: (
+  clientCredentials: { clientId: string; clientSecret: string },
+  idc: { id: string; region: string }
+) => OidcLoginSteps<AWSAuthorizeResponse, AWSTokenResponse> = (
+  clientCredentials: { clientId: string; clientSecret: string },
+  idc: { id: string; region: string }
+) => ({
+  authorize: async () => {
+    const authorizeResponse = await authorize<AWSAuthorizeResponse>(
+      idcAuthorizeRequestBuilder(clientCredentials, idc)
+    );
+    print2(`Please use the opened browser window to continue your idc device authorization.
+  
+      When prompted, confirm that aws idc page displays this code:
+      
+        ${authorizeResponse.userCode}
+      
+      Waiting for authorization...
+      `);
+    void open(authorizeResponse.verificationUriComplete);
```

Comment:
do we need `void` here? 🤔

Reply:
we do not need evaluate the result of the promise to open the browser window.

Reply:
@GGonryun we use `void` to indicate explicitly that we are not handling the promise; errors in the promise will crash the program.